c++/src/kj/compat/http.c++

```diff
@@ -4093,6 +4093,103 @@ private:
   kj::ArrayPtr<byte> leftover;
 };
 
+class AsyncIoStreamWithDeferredInput final: public kj::AsyncIoStream {
+  // This AsyncIoStream immediately allows write to the given wrapped AsyncIoStream
+  // but will delay reads until after a guard promise is resolved. When the guard
+  // promise resolves, it may provide a released buffer that will be read out first.
+  // The primary use case for this impl is to support pipelined CONNECT calls which
+  // optimistically allow outbound writes to happen while establishing the CONNECT
+  // tunnel has not yet been completed. If the guard promise rejects, the stream
+  // is permanently errored and existing pending calls (reads and writes) are canceled.
+public:
+  AsyncIoStreamWithDeferredInput(
+      kj::Own<kj::AsyncIoStream> inner,
+      kj::Promise<HttpInputStreamImpl::ReleasedBuffer> guard)
+      : inner(kj::mv(inner)),
+        guard(handleGuard(kj::mv(guard))) {}
+
+  // AsyncInputStream
+  Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override {
+    KJ_IF_MAYBE(exception, permanentException) {
+      return kj::cp(*exception);
+    }
+    return guard.addBranch().then([this, buffer, minBytes, maxBytes] {
```

Comment:
Can you follow the pattern of `PromisedAsyncIoStream`, which avoids the overhead of `addBranch()` in the case that the promise has already resolved?